### PR TITLE
adding orphaned snapshot delete support

### PR DESCRIPTION
**Pull request by ivanhorvath**

### CODE REVIEW

**Comment by ivanhorvath** [comment]
[test]

**Comment by ivanhorvath** [comment]
[test]

File: openshift_tools/cloud/aws/ebs_snapshotter.py [hunk 1]
```diff
@@ -324,42 +330,55 @@ def get_expired_snapshots(snapshots, hourly_backups, daily_backups, weekly_backu
...
+        # sort the snapshots into the following categories:
+        # - snapshots with a volume that currently exists
+        # - snapshots that are orphaned and have no volume anymore
+        # - ignore orphaned snapshots that we did not create, that have no 'snapshot' tag
+        for snapshot in all_snapshots:
+            if snapshot.volume_id in volume_ids:
+                volume_snapshots[snapshot.volume_id].append(snapshot)
+            else:
+                if 'snapshot' in snapshot.tags and snapshot.tags['snapshot'] in SUPPORTED_SCHEDULES:
+                    orphan_snapshots.append(snapshot)
```
Comment (twiest) [inline openshift_tools/cloud/aws/ebs_snapshotter.py]: Consider adding an early exit for the case where volumes and snapshots are 0:

```python
       if not volumes and not orphan_snapshots:
            # Early exit for any region that doesn't have snapshot-able volumes or orphaned snapshots.
            return (all_expired_snapshots, deleted_snapshots, errors)
```

File: openshift_tools/cloud/aws/ebs_snapshotter.py [hunk 2]
```diff
@@ -396,9 +416,29 @@ def trim_snapshots(self, hourly_backups, daily_backups, weekly_backups, monthly_
...
-        self.verbose_print("Number of snapshot deletion errors: %d" % len(errors), prefix="  ")
+        # delete all orphaned snapshots older than N days
+        self.verbose_print("Removing orphaned snapshots older than {} days.\n".format(delete_orphans_older_than),
+                           prefix=" " * 2)
+
+        cutoff_date = datetime.utcnow() - timedelta(days=delete_orphans_older_than)
+        orphan_delete_counter = 0
+        for snapshot in orphan_snapshots:
+            snap_date = datetime.strptime(snapshot.start_time, '%Y-%m-%dT%H:%M:%S.000Z')
+            if snap_date < cutoff_date and 'snapshot' in snapshot.tags:
```
Comment (twiest) [inline openshift_tools/cloud/aws/ebs_snapshotter.py]: Consider removing ```'snapshot' in snapshot.tags``` as we check above that it's in there.

File: openshift_tools/cloud/aws/ebs_snapshotter.py [hunk 3]
```diff
@@ -324,42 +330,55 @@ def get_expired_snapshots(snapshots, hourly_backups, daily_backups, weekly_backu
...
-        all_snapshots = self.ec2.get_all_snapshots()
+        # sort the snapshots into the following categories:
+        # - snapshots with a volume that currently exists
+        # - snapshots that are orphaned and have no volume anymore
+        # - ignore orphaned snapshots that we did not create, that have no 'snapshot' tag
+        for snapshot in all_snapshots:
+            if snapshot.volume_id in volume_ids:
+                volume_snapshots[snapshot.volume_id].append(snapshot)
+            else:
+                if 'snapshot' in snapshot.tags and snapshot.tags['snapshot'] in SUPPORTED_SCHEDULES:
```
Comment (twiest) [inline openshift_tools/cloud/aws/ebs_snapshotter.py]: Add a comment to not remove ```snapshot.tags['snapshot'] in SUPPORTED_SCHEDULES``` as it's crucial for this to only remove snapshots that are part of this snapshotting / trimming system.

File: openshift_tools/cloud/aws/ebs_snapshotter.py [hunk 4]
```diff
@@ -396,9 +416,29 @@ def trim_snapshots(self, hourly_backups, daily_backups, weekly_backups, monthly_
...
+        for snapshot in orphan_snapshots:
+            snap_date = datetime.strptime(snapshot.start_time, '%Y-%m-%dT%H:%M:%S.000Z')
+            if snap_date < cutoff_date and 'snapshot' in snapshot.tags:
+                try:
+                    if dry_run:
+                        self.print_dry_run_msg(prefix=" " * 10, end=' Would delete {}\n'.format(snapshot.tags['Name']))
+                    else:
+                        snapshot.delete()
+                        orphan_delete_counter += 1
+                # pylint: disable=broad-except
```
Comment (twiest) [inline openshift_tools/cloud/aws/ebs_snapshotter.py]: Please add a reason block here per our best practices:
https://github.com/openshift/openshift-tools/blob/prod/docs/best_practices_guide.adoc#All-PyLint-rule-disables-MUST-be-documented-in-the-code

File: openshift_tools/cloud/aws/ebs_snapshotter.py [hunk 5]
```diff
@@ -151,6 +154,7 @@ def create_snapshots(self, schedule, script_name=None, dry_run=False, sleep_betw
             # Status: permanently disabled
             # pylint: disable=broad-except
             except Exception as ex:
+                # pylint: disable=no-member
```
Comment (twiest) [inline openshift_tools/cloud/aws/ebs_snapshotter.py]: Please add a reason block here per our best practices:
https://github.com/openshift/openshift-tools/blob/prod/docs/best_practices_guide.adoc#All-PyLint-rule-disables-MUST-be-documented-in-the-code

File: openshift_tools/cloud/aws/ebs_snapshotter.py [hunk 6]
```diff
@@ -272,6 +276,7 @@ def get_expired_snapshots(snapshots, hourly_backups, daily_backups, weekly_backu
         temp = []
 
         for t in target_backup_times:
+            # pylint: disable=singleton-comparison
```
Comment (twiest) [inline openshift_tools/cloud/aws/ebs_snapshotter.py]: Please add a reason block here per our best practices:
https://github.com/openshift/openshift-tools/blob/prod/docs/best_practices_guide.adoc#All-PyLint-rule-disables-MUST-be-documented-in-the-code

File: openshift_tools/cloud/aws/ebs_snapshotter.py [hunk 7]
```diff
@@ -301,6 +306,7 @@ def get_expired_snapshots(snapshots, hourly_backups, daily_backups, weekly_backu
                     #date the date ranges and the snapshots
                     # are sorted chronologically, we know this
                     #snapshot isn't in an earlier date range):
+                    # pylint: disable=singleton-comparison
```
Comment (twiest) [inline openshift_tools/cloud/aws/ebs_snapshotter.py]: Please add a reason block here per our best practices:
https://github.com/openshift/openshift-tools/blob/prod/docs/best_practices_guide.adoc#All-PyLint-rule-disables-MUST-be-documented-in-the-code

File: openshift_tools/cloud/aws/ebs_snapshotter.py [hunk 8]
```diff
@@ -324,42 +330,55 @@ def get_expired_snapshots(snapshots, hourly_backups, daily_backups, weekly_backu
...
-        # Get this volume's specific snapshots
-        vol_snaps = [s for s in all_snapshots if s.volume_id == volume.id]
-        EbsSnapshotter.sort_snapshots(vol_snaps)
-        return vol_snaps
 
     # Reason: disable pylint too-many-arguments because this is the API I want to expose.
     # Status: permanently disabled
     # pylint: disable=too-many-arguments
-    def trim_snapshots(self, hourly_backups, daily_backups, weekly_backups, monthly_backups, dry_run=False):
+    # pylint: disable=too-many-statements
```
Comment (twiest) [inline openshift_tools/cloud/aws/ebs_snapshotter.py]: Please add a reason block here per our best practices:
https://github.com/openshift/openshift-tools/blob/prod/docs/best_practices_guide.adoc#All-PyLint-rule-disables-MUST-be-documented-in-the-code

File: openshift_tools/cloud/aws/ebs_snapshotter.py [hunk 9]
```diff
@@ -370,24 +389,25 @@ def trim_snapshots(self, hourly_backups, daily_backups, weekly_backups, monthly_
...
-                        self.print_dry_run_msg(prefix="          ")
+                        self.print_dry_run_msg(prefix=" " * 10)
                     else:
                         exp_snap.delete()
                         deleted_snapshots.append(exp_snap)
                 # Reason: disable pylint broad-except because we want to continue on error.
                 # Status: permanently disabled
                 # pylint: disable=broad-except
                 except Exception as ex:
+                    # pylint: disable=no-member
```
Comment (twiest) [inline openshift_tools/cloud/aws/ebs_snapshotter.py]: Please add a reason block here per our best practices:
https://github.com/openshift/openshift-tools/blob/prod/docs/best_practices_guide.adoc#All-PyLint-rule-disables-MUST-be-documented-in-the-code

File: openshift_tools/cloud/aws/ebs_snapshotter.py [hunk 10]
```diff
@@ -324,42 +330,55 @@ def get_expired_snapshots(snapshots, hourly_backups, daily_backups, weekly_backu
...
     # Status: permanently disabled
     # pylint: disable=too-many-arguments
-    def trim_snapshots(self, hourly_backups, daily_backups, weekly_backups, monthly_backups, dry_run=False):
+    # pylint: disable=too-many-statements
+    def trim_snapshots(self, hourly_backups,
+                       daily_backups,
+                       weekly_backups,
+                       monthly_backups,
+                       delete_orphans_older_than,
+                       dry_run=False):
```
Comment (twiest) [inline openshift_tools/cloud/aws/ebs_snapshotter.py]: This method is now very big. Consider spending some time breaking out some of its functionality into separate methods.

File: openshift_tools/cloud/aws/ebs_snapshotter.py [hunk 11]
```diff
@@ -370,24 +439,28 @@ def trim_snapshots(self, hourly_backups, daily_backups, weekly_backups, monthly_
...
                     else:
                         exp_snap.delete()
                         deleted_snapshots.append(exp_snap)
                 # Reason: disable pylint broad-except because we want to continue on error.
                 # Status: permanently disabled
                 # pylint: disable=broad-except
                 except Exception as ex:
+                    # Reason: disable pylint no-member because this was acceptable before,
+                    # pylint got stricter, but we do not want to modify this code currently
+                    # Statue: permanently disabled
```
Comment (twiest) [inline openshift_tools/cloud/aws/ebs_snapshotter.py]: fix Statue

**Review by twiest — CHANGES_REQUESTED** [review]

**Review by twiest — APPROVED** [review]
Looks great!   :+1:

**Pull request merged**